### PR TITLE
IGNITE-22772 Sql. Fix row type derivation for renaming in FROM clause

### PR DESCRIPTION
**Pull request by korlov42**
https://issues.apache.org/jira/browse/IGNITE-22772

This patch introduces proper handling of system/hidden columns during column renaming in `FROM` clause (`SELECT * FROM my_table AS new_name(new_c1, new_c2, ...)`). Provided solution has one limitation: it's impossible to rename system columns directly. But this limitation has workaround: table may be replaced with sub-select explicitly mentioning system columns in SELECT clause (see last test case in `rename_columns_in_from.test`). 

-------------
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/TableDescriptor.java [hunk 1]
```diff
@@ -41,6 +41,9 @@ public interface TableDescriptor extends InitializerExpressionFactory, Iterable<
      */
     RelDataType rowType(IgniteTypeFactory factory, @Nullable ImmutableBitSet usedColumns);
 
+    /** Returns type of the row excluding all {@link ColumnDescriptor#hidden() hidden} columns. */
+    RelDataType rowTypeSansHidden();
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/TableDescriptor.java:45]: Why `Sans`? What does it mean?

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/TableDescriptor.java:45]: this means "without" or "ignoring" ([link](https://dictionary.cambridge.org/dictionary/english/sans) to dictionary). This term is used in core calcite in a few places:
* `SqlTypeUtil#equalSansNullability(org.apache.calcite.rel.type.RelDataType, org.apache.calcite.rel.type.RelDataType)`
* `RelTraitSet#isDefaultSansConvention`
* `RelDataType#equalsSansFieldNames`
* `SqlValidatorNamespace#getRowTypeSansSystemColumns`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+                                    return Pair.of(name, f.getType());
+                                }
+                        )
+                )
+                .kind(rowType.getStructKind())
+                .build();
+
+        return validator.getTypeFactory()
+                .createTypeWithNullability(aliasedType, rowType.isNullable());
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java:139]: I understand that this is a copy-paste from the calcite, but just curious what is the difference between
```
return createTypeWithNullability(aliasedType, rowType.isNullable());
```
and
```
return aliasedTypeBuilder...nullable(rowType.isNullable()).build();
```

**Review by xtern — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java:139]:
```
    /**
     * Changes the nullability of the last field added.
     *
     * @throws java.lang.IndexOutOfBoundsException if no fields have been
     *                                             added
     */
    public Builder nullable(boolean nullable) {
      RelDataType lastType = types.get(types.size() - 1);
      if (lastType.isNullable() != nullable) {
        final RelDataType type =
            typeFactory.createTypeWithNullability(lastType, nullable);
        types.set(types.size() - 1, type);
      }
      return this;
    }
```

it changes nullability for the last column, but we need to make an entire row nullable

Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java:139]: :woman_facepalming:  thanks

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        // Although this may look like proper way to address `system` columns, in fact it's not:
+        //     1) Calcite has different notion of `system` columns. First of all, all implementations
+        //     of namespace just propagate call to getRowType() method. Second, system columns are
+        //     registered only when converter meets LogicalJoin.
+        //
+        //     2) Row type returned from this method will be used to resolve columns. Thus, if we skip
+        //     `system` columns validator won't be able to resolve any reference of such columns (given
+        //     that they are registered only in case of joins)
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java]: fixed, thanks

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                            getString(rowTypeSansHidden), columnNames.size()));
+        }
+
+        // Although this may look like proper way to address `system` columns, in fact it's not:
+        //     1) Calcite has different notion of `system` columns. First of all, all implementations
+        //     of namespace just propagate call to getRowType() method. Second, system columns are
+        //     registered only when converter meets LogicalJoin.
+        //
+        //     2) Row type returned from this method will be used to resolve columns. Thus, if we skip
+        //     `system` columns validator won't be able to resolve any reference of such columns (given
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteAliasNamespace.java]: fixed, thanks

**Review by ygerzhedovich — APPROVED** [review]

**Pull request merged**